### PR TITLE
framework/araui: Enhancement of UV mapping quality

### PR DESCRIPTION
**Pull request by gcjjyy**
Before, uv mapping is based on the basic algorithm.
Enhanced algorithm is applied.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [3876692eee42278903b9f5f36d4892107d055b01] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [3876692eee42278903b9f5f36d4892107d055b01] - Code Rule Check Result:
framework/include/araui/ui_commons.h:181: ERROR: [SPC_M_KWD] missing space after enum definition

**Comment by seinfra** [comment]
Target : [71c5ab0ad52a5647263d1c24556268f3fafa7d65] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [71c5ab0ad52a5647263d1c24556268f3fafa7d65] - Code Rule Check OK.

File: framework/src/araui/renderer/ui_renderer.c [hunk 1]
```diff
@@ -33,15 +34,14 @@
 #define MAX_RENDERER_MATRIX_STACK (256)
 #define UI_TM (g_rc.tm_stack[g_rc.sp])
 
+#define SUB_DIVIDE_SHIFT (4)
```
Comment (tyha1115) [inline framework/src/araui/renderer/ui_renderer.c]: please add `UI_` prefix

File: framework/src/araui/renderer/ui_renderer.c [hunk 2]
```diff
@@ -33,15 +34,14 @@
...
 /****************************************************************************
  * Private function declaration
  ****************************************************************************/
-static void _ui_fill_bottom_flat_triangle_uv(
-	ui_coord_t coord1, ui_coord_t coord2, ui_coord_t coord3,
-	ui_uv_t uv1, ui_uv_t uv2, ui_uv_t uv3);
-static void _ui_fill_top_flat_triangle_uv(
-	ui_coord_t coord1, ui_coord_t coord2, ui_coord_t coord3,
-	ui_uv_t uv1, ui_uv_t uv2, ui_uv_t uv3);
+static void DrawSegment(int32_t y1, int32_t y2);
```
Comment (tyha1115) [inline framework/src/araui/renderer/ui_renderer.c]: please add `_ui_` prefix for static function and do not use capital letters.

File: framework/src/araui/renderer/ui_renderer.c [hunk 3]
```diff
@@ -67,6 +67,13 @@ ui_render_context_t g_rc = {
 	.tex_pf = UI_PIXEL_FORMAT_UNKNOWN
 };
 
+float g_left_dxdy, g_right_dxdy, g_leftx, g_rightx;
```
Comment (tyha1115) [inline framework/src/araui/renderer/ui_renderer.c]: please declare variables in different lines

File: framework/src/araui/renderer/ui_renderer.c [hunk 4]
```diff
@@ -188,216 +195,313 @@ void ui_renderer_set_texture(uint8_t *bitmap, int32_t width, int32_t height, ui_
...
-		tex_color.r = g_rc.texture[offset];
-		tex_color.g = g_rc.texture[offset + 1];
-		tex_color.b = g_rc.texture[offset + 2];
-		tex_color.a = g_rc.texture[offset + 3];
-		//ui_dal_put_pixel_rgba8888(x + i, y, *(ui_color_t *)&tex_color);
-		ui_dal_put_pixel_rgba8888(x + i, y, 0xff0000ff);
-
-		u += u_step;
-		v += v_step;
+	float u_a, v_a, z_a;
```
Comment (tyha1115) [inline framework/src/araui/renderer/ui_renderer.c]: same as above.

File: framework/src/araui/renderer/ui_renderer.c [hunk 5]
```diff
@@ -188,216 +195,313 @@ void ui_renderer_set_texture(uint8_t *bitmap, int32_t width, int32_t height, ui_
...
-		uv[1].v = UI_GET_WEIGHTED_VALUE(uv2.v, uv3.v, weight);
-
-		if (curx2 >= curx1) {
-			x = (int32_t)curx1;
-			len = (int32_t)curx2 - x + 1;
-		} else {
-			x = (int32_t)curx2;
-			len = (int32_t)curx1 - x + 1;
-			UI_SWAP(uv[0], uv[1]);
+	float u, v, z, Z;
```
Comment (tyha1115) [inline framework/src/araui/renderer/ui_renderer.c]: same as above.

File: framework/src/araui/widgets/ui_image_widget.c [hunk 6]
```diff
@@ -112,7 +112,7 @@ static void _ui_image_widget_set_image_func(void *userdata)
 static void ui_image_widget_render_func(ui_widget_t widget, uint32_t dt)
 {
 	ui_image_widget_body_t *body;
-	ui_coord_t coord[4];
+	ui_vec3_t v1, v2, v3, v4;
```
Comment (tyha1115) [inline framework/src/araui/widgets/ui_image_widget.c]: same as above.

**Comment by seinfra** [comment]
Target : [fc26a82315f9e21cced828a79f44cc8b4bae9b60] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fc26a82315f9e21cced828a79f44cc8b4bae9b60] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [98e76820fe3ce03030ec114ba23f6704ed67455c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [98e76820fe3ce03030ec114ba23f6704ed67455c] - Code Rule Check OK.

**Pull request merged**